examples/lwip_ethernet_tap/main.c

```diff
@@ -0,0 +1,438 @@
+/* vim: set sw=2 expandtab tw=80: */
+
+#include <libtock/services/alarm.h>
+#include <libtock/tock.h>
+#include <libtock/util/streaming_process_slice.h>
+#include <lwip/apps/httpd.h>
+#include <lwip/dhcp.h>
+#include <lwip/err.h>
+#include <lwip/init.h>
+#include <lwip/netif.h>
+#include <lwip/prot/ethernet.h>
+#include <lwip/sys.h>
+#include <lwip/timeouts.h>
+#include <stdio.h>
+#include <string.h>
+
+// ########## Settings
+
+// Set this variable to `true` to enable debug print messages. These may slow
+// down the application significantly.
+#define DEBUG false
+
+// Control whether the application shall provide a buffer for additional
+// feedback on transmitted frames (e.g. their timestamps).
+#define TX_FRAME_INFO_ENABLE 1
+
+// Maximum size of a frame which can be transmitted with the
+// `EthernetTapDriver`. Currently hard-coded to `1522 - 4` bytes, for
+// an Ethernet frame with an 802.1q VLAN tag with a 1500 byte payload
+// MTU, excluding the 4-byte FCS.
+#define ETHERNET_MTU 1518
+
+// How much space to allocate for frame receive buffers (in units of
+// `ETHERNET_MTU + TAP_DRIVER_FRAME_HEADER_LEN`, at least 2):
+#define RX_FRAME_BUFFER_FRAMES 8
+
+// ########## Misc macros and constants
+
+const uint32_t tap_driver_num = 0x90010;
+
+#define debug_print(...)   \
+  do {                     \
+    if (DEBUG)             \
+    printf(__VA_ARGS__); \
+  } while (0)
+
+// Transmission metadata feedback buffer len:
+#if TX_FRAME_INFO_ENABLE
+#define TAP_DRIVER_TX_FRAME_INFO_LEN 8
+#endif
+
+// Frame header prepended by the `EthernetTapDriver`:
+#define TAP_DRIVER_FRAME_HEADER_LEN 12
+
+// ########## Global buffers
+
+// Buffer to transmit frames from / to the kernel:
+static uint8_t tx_frame_buffer[ETHERNET_MTU];
+
+// This buffer is used for the RX_FRAMES streaming process slice.
+#define RX_FRAME_BUFFER_LEN \
+  (ETHERNET_MTU + TAP_DRIVER_FRAME_HEADER_LEN) * RX_FRAME_BUFFER_FRAMES
+static uint8_t rx_frames_streaming_buffer[RX_FRAME_BUFFER_LEN];
```

Comment:
Does this account for the extra 8 bytes *2 that the StreamingProcessSlice needs? Otherwise this holds one packet less than you were expecting, right? Actually, two less, I think, since we split it in half.

Reply:
That's correct, and it does not! Great point, will fix and add the `#define` as you recommended in #492.